dnf5/commands/autoremove/autoremove.cpp

```diff
@@ -0,0 +1,57 @@
+/*
+Copyright Contributors to the libdnf project.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "autoremove.hpp"
+
+#include <libdnf/rpm/package_query.hpp>
+
+namespace dnf5 {
+
+using namespace libdnf::cli;
+
+void AutoremoveCommand::set_parent_command() {
+    auto * arg_parser_parent_cmd = get_session().get_argument_parser().get_root_command();
+    auto * arg_parser_this_cmd = get_argument_parser_command();
+    arg_parser_parent_cmd->register_command(arg_parser_this_cmd);
+    arg_parser_parent_cmd->get_group("software_management_commands").register_argument(arg_parser_this_cmd);
+}
+
+void AutoremoveCommand::set_argument_parser() {
+    get_argument_parser_command()->set_description(
+        "Remove all unneeded packages originally installed as dependencies.");
+}
+
+void AutoremoveCommand::configure() {
+    auto & context = get_context();
+    context.set_load_system_repo(true);
+    context.set_load_available_repos(Context::LoadAvailableRepos::NONE);
+}
+
+void AutoremoveCommand::run() {
+    auto & ctx = get_context();
+    libdnf::rpm::PackageQuery unneeded(ctx.base);
+    unneeded.filter_unneeded();
+    auto goal = get_context().get_goal();
+    goal->set_allow_erasing(true);
```

Comment:
I am curious, why the command requires `allow_erasing == true`?

I am suggesting that it is a left over from handling of spec code from DNF4. Or did I overlook something?
In principle unneeded packages are not supposed to remove any dependent package.
The question is how it works with excludes when I exclude top package from dependency tree.

Reply:
I was thinking about a chain of dependencies:
```
# dnf5 install Pkg
[Pkg - userinstalled] ---requires--> [Unneeded - dependency] ---requires--> [UnneededDep]

# dnf5 upgrade Pkg && new version of Pkg dropped dependency on Unneeded

# dnf5 autoremove
```

Here hopefully `Unneeded` package should be removed with all it's dependencies (`UnneededDep`). I think to enable removal of unused dependencies this flag is needed. Or am I mistaken?

Reply:
From DNF4:

```
        if any([self.opts.grp_specs, self.opts.pkg_specs, self.opts.filenames]):
            self.base.conf.clean_requirements_on_remove = True
            demands.allow_erasing = True
            # disable all available repos to delete whole dependency tree
            # instead of replacing removable package with available packages
            demands.available_repos = False
        else:
            demands.available_repos = True
            demands.fresh_metadata = False
```
As you can see, `demands.allow_erasing = True` is only set when argument (spec) is provided.
I believe that line `goal->set_allow_erasing(true);` could be removed without any issue. I also believe that unneeded query does not return only unneeded leave packages but also unneeded dependencies of such a packages.

Reply:
OK, I've done some tests and it seems you are right :) I've removed the flag.